compression_verifier.go

```diff
@@ -0,0 +1,238 @@
+)
+
+type (
+	columnCompressionConfig map[string]string
+
+	// TableColumnCompressionConfig represents compression configuration for a
+	// column in a table as table -> column -> compression-type
+	// ex: books -> contents -> snappy
+	TableColumnCompressionConfig map[string]columnCompressionConfig
+)
+
+// UnsupportedCompressionError is used to identify errors resulting
+// from attempting to decompress unsupported algorithms
+type UnsupportedCompressionError struct {
+	table     string
+	column    string
+	algorithm string
+}
+
+func (e UnsupportedCompressionError) Error() string {
+	return "Compression algorithm: " + e.algorithm +
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides support for verifying the payload of compressed columns that
+// may have different hashes for the same data by first decompressing the compressed
+// data before fingerprinting
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	supportedAlgorithms     map[string]struct{}
+	tableColumnCompressions TableColumnCompressionConfig
+}
+
+// GetCompressedHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The GetCompressedHashes method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used to verify equality
+func (c *CompressionVerifier) GetCompressedHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (map[uint64][]byte, error) {
+	c.logger.WithFields(logrus.Fields{
+		"tag":   "compression_verifier",
+		"table": table,
+	}).Info("decompressing table data before verification")
+
+	tableCompression := c.tableColumnCompressions[table]
+
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	resultSet := make(map[uint64][]byte)
+	for rows.Next() {
+		rowData, err := ScanByteRow(rows, len(columns)+1)
+		if err != nil {
+			return nil, err
+		}
+
+		pk, err := strconv.ParseUint(string(rowData[0]), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decompress the applicable columns and then hash them together
+		// to create a fingerprint. decompressedRowData contains a map of all
+		// the non-compressed columns and associated decompressed values by the
+		// index of the column
+		decompressedRowData := [][]byte{}
+		for idx, column := range columns {
+			if algorithm, ok := tableCompression[column.Name]; ok {
+				// rowData contains the result of "SELECT pkColumn, * FROM ...", so idx+1 to get each column
+				decompressedColData, err := c.Decompress(table, column.Name, algorithm, rowData[idx+1])
```

Comment:
same question here. why the `+1`? Is that because `rowData` contains the primary key column but `columns` doesn't?

Reply:
nvm I just saw the comment 🤦‍♂️